jujugui/static/gui/src/app/state/state.js

```diff
@@ -0,0 +1,323 @@
+      state = this._parseSearch(parts.splice(1), state);
+      // If we have a search path in the URL then we can ignore everything else.
+      return {error, state};
+    }
+    // Working backwards to split up the URL.
+    const guiIndex = parts.indexOf(PATH_DELIMETERS.get('gui'));
+    if (guiIndex > -1) {
+      ({state, error} = this._parseGUI(parts.splice(guiIndex), state));
+      if (error !== null) {
+        return {error, state};
+      }
+    }
+    // Extract out the user sections.
+    if (parts.includes(PATH_DELIMETERS.get('user'))) {
+      ({state, parts, error} = this._parseUser(parts, state));
+      if (error !== null) {
+        return {error, state};
+      }
+    }
+    // By this point there should only be the 'store only' content left.
+    if (!state.store && parts.length) {
+      // If there are more than 3 parts then this is an invalid url.
+      if (parts.length > 3) {
+        error = 'invalid store path.';
+      } else {
+        state.store = parts.join('/');
+      }
+      // If we have more content here but there is already a store populated.
+      // then this is an invalid url.
+    } else if (state.store && parts.length) {
+      error = 'invalid store path.';state.store = parts.join('/');
+    }
+    return {error, state};
+  }
+
+  /**
+    Inspects the URL path to see if it is for the root.
+    @param {Array} urlParts - The URL path split into parts.
+    @param {Object} state - The application state object as being parsed
+      from the URL.
+    @return {Object} The updated state to contain the root value, if any.
+  */
+  _parseRoot(urlParts, state) {
+    ROOT_RESERVED.some(key => {
+      if (urlParts[0] === key) {
+        state.root = key;
+        return true;
+      }
+    });
+    return state;
+  }
+
+  /**
+    Parses the search portion of the URL without the
+    PATH_DELIMETERS.get('search') key value.
+    @param {Array} urlParts - The URL path split into parts.
+    @param {Object} state - The application state object as being parsed
+      from the URL.
+    @return {Object} The updated state to contain the search value, if any.
+  */
+  _parseSearch(urlParts, state) {
+    if (urlParts.length > 0) {
+      state.search = urlParts.join('/');
+    }
+    return state;
+  }
+
+  /**
+    Parses the GUI portion of the URL without the
+    PATH_DELIMETERS.get('gui') key value.
+    @param {Array} urlParts - The URL path split into parts.
+    @param {Object} state - The application state object as being parsed
+      from the URL.
+    @return {Object} The updated state to contain the search value, if any.
+  */
+  _parseGUI(urlParts, state) {
+    let error = null;
+    let indexes = [];
+    GUI_PATH_DELIMETERS.forEach(section => {
+      const index = urlParts.indexOf(section);
+      if (index > -1) {
+        indexes.push(index);
+      }
+    });
+    // If there were no sections found then this is an invalid URL.
+    if (!indexes.length) {
+      error = 'invalid GUI path.';
```

Comment:
I'd avoid this kind of punctation on all errors. That allows for easy concatenation, for instance:
```javascript
({error, state} = parseFoo());
if (error !== nil) {
  error = 'cannot parse foo: ' + error;
  return {error, state};
}
```